lib/widgets/amount_form_field.dart

```diff
@@ -99,3 +99,37 @@ class AmountFormField extends TextFormField {
     };
   }
 }
+
+class SatAmountFormFieldFormatter extends TextInputFormatter {
+  final RegExp _pattern = RegExp(r'[^\d*]');
+
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue,
+    TextEditingValue newValue,
+  ) {
+    final raw = newValue.text.replaceAll(_pattern, '');
+    if (raw.isEmpty) {
+      return newValue.copyWith(
+        text: '',
+        selection: TextSelection.collapsed(offset: 0),
+      );
+    }
+
+    var value;
+    try {
+      value = Int64.parseInt(raw.length > 18 ? raw.substring(0, 18) : raw);
+    } catch (ignored) {
+      value = Int64(0);
+    }
+
+    final formatted = Currency.SAT.format(
+      value,
+      includeDisplayName: false,
+      includeCurrencySymbol: false,
+    );
+    return newValue.copyWith(
```

Comment:
As it is right now the cursor always jumps to the end of the text even if the user adds a digit in the middle.
I think it is better to preserve the selection and given the fact that a space can be added or removed I think something similar to the following suggestion may work:

```suggestion
    var diff = formatted.length - oldValue.text.length;
    var newOffset = newValue.selection.start;
    if (diff > 1) {
      newOffset += 1;
    }
    if (diff < -1) {
      newOffset -= 1;
    }
    return newValue.copyWith(
      text: formatted,
      selection: TextSelection.collapsed(offset: newOffset),
    );
```

Reply:
Good catch, the code you suggested work like a charm, I had to add a small check only `if (formatted != oldValue.text)…` and use the old selection as else statement `newOffset = oldValue.selection.start;`. that is to avoid a wrong behaviour when you try to insert an invalid character.

Reply:
btw How it looks like:

https://user-images.githubusercontent.com/1225438/122784577-f6077080-d288-11eb-9a0c-35640be3ef67.mp4